src/Microsoft.ML.Transforms/GcnTransform.cs

```diff
@@ -683,19 +737,153 @@ public static CommonOutputs.TransformOutput Normalize(IHostEnvironment env, LpNo
             XmlInclude = new[] { @"<include file='../Microsoft.ML.Transforms/doc.xml' path='doc/members/member[@name=""GcNormalize""]/*' />" })]
-        public static CommonOutputs.TransformOutput GcNormalize(IHostEnvironment env, LpNormNormalizerTransform.GcnArguments input)
+        public static CommonOutputs.TransformOutput GcNormalize(IHostEnvironment env, LpNormalizingTransformer.GcnArguments input)
         {
             var h = EntryPointUtils.CheckArgsAndCreateHost(env, "GcNormalize", input);
-            var xf = new LpNormNormalizerTransform(h, input, input.Data);
+            var xf = LpNormalizingTransformer.Create(h, input, input.Data);
             return new CommonOutputs.TransformOutput()
             {
                 Model = new TransformModel(h, xf, input.Data),
                 OutputData = xf
             };
         }
     }
+
+    /// <summary>
+    /// Base estimator class for LpNorm and Gcn normalizers.
+    /// </summary>
+    public abstract class LpNormalizingEstimatorBase : TrivialEstimator<LpNormalizingTransformer>
+    {
+        /// <summary>
+        /// The kind of unit norm vectors are rescaled to. This enumeration is serialized.
+        /// </summary>
+        public enum NormalizerKind : byte
+        {
+            L2Norm = 0,
+            StdDev = 1,
+            L1Norm = 2,
+            LInf = 3
+        }
+
+        internal static class Defaults
+        {
+            public const NormalizerKind NormKind = NormalizerKind.L2Norm;
+            public const bool LpSubMean = false;
+            public const bool GcnSubMean = true;
+            public const bool UseStdDev = false;
+            public const float Scale = 1;
+        }
+
+        /// <summary>
+        /// Create a <see cref="LpNormalizingEstimatorBase"/> that takes multiple pairs of columns.
+        /// </summary>
+        public LpNormalizingEstimatorBase(IHostEnvironment env, params LpNormalizingTransformer.ColumnInfoBase[] columns)
+            : base(Contracts.CheckRef(env, nameof(env)).Register(nameof(LpNormalizingEstimator)), new LpNormalizingTransformer(env, columns))
+        {
+
+        }
+
+        public override SchemaShape GetOutputSchema(SchemaShape inputSchema)
+        {
+            Host.CheckValue(inputSchema, nameof(inputSchema));
+            var result = inputSchema.Columns.ToDictionary(x => x.Name);
+            foreach (var colPair in Transformer.Columns)
+            {
+                if (!inputSchema.TryFindColumn(colPair.Input, out var col))
+                    throw Host.ExceptSchemaMismatch(nameof(inputSchema), "input", colPair.Input);
+                var reason = LpNormalizingTransformer.TestColumn(col.ItemType);
+                if (reason != null)
+                    throw Host.ExceptUserArg(nameof(inputSchema), reason);
+                if (col.Kind != SchemaShape.Column.VectorKind.Vector)
+                    throw Host.ExceptSchemaMismatch(nameof(inputSchema), "input", colPair.Input);
```

Comment:
>ExceptSchemaMismatch [](start = 31, length = 20)

wrong message here: you'll complain that the column is not found, but you should complain that the type is incorrect